### DISCUSSION TITLE
Looking for feature requests related to small matrices

### DISCUSSION THREAD

**Discussion by FantasyVR** [taichi-dev/taichi#5103] [Ideas]
In many engineering fields, we need to use functions related to small matrices (2x2 or 3x3) such as single value decomposition (SVD), eigendecomposition, and more. Some are already implemented in Taichi shown in the following table.  

| Vector          | Matrix                  |
|-----------------|-------------------------|
| V.normalized()  | ti.polar_decompose(M)    |
| V.norm()        | ti.svd(M)                |
| V.norm_inv()    | ti.sym_eig(M)            |
| V.norm_sqr()    | ti.eig(M)                |
| V.sum()         | ti.solve(M, b)              |
| V1.dot(V2)      | M.norm()                |
| V1.cross(V2)    | M.norm_inv()            |
|                 | M.norm_sqr()            |
|                 | M.sum()                 |
|                 | M.transpose()           |
|                 | M.determinant()         |
|                 | M.trace()               |

**Note** : V for vector, M for matrix

But we feel that there may be functions that are needed in areas we are unaware of. In that spirit, we'd like to hear more from the community. 

If you ever tried to implement a feature by yourself, chances are high that many people in the community also need it.  We encourage you to bring up relevant features to this discussion. Together, we will implement these wanted small linear algebra functions to make our Taichi kernels easier to write.

**Comment by ghost** [comment]
Any recommendations for books for teaching how to write code of matrix from scratch without calling exsited libary

**Reply by FantasyVR** [reply]
Hi @eatcosmos , I think this [doc](https://arxiv.org/pdf/2201.00145.pdf) is good to learn.

**Reply by EnochZhu** [reply]
> 

nice doc

**Comment by EnochZhu** [comment]
V.prod() or M.prod() for the product of array elements over.

**Reply by EnochZhu** [reply]
Hi,please consider extending 'taichi.math.reflect()' to Matrix . This would be helpful to treat eigenvectors matrix.

**Comment by tengzhang48** [comment]
It will be nice to be able to calculate Logarithm of a matrix.

**Comment by kurt-rhee** [comment]
I would like to be able to use quaternions.  I would like to go back and froth between quaternions and matrix rotations

**Comment by kurt-rhee** [comment]
I would also like a projection matrix

**Reply by kurt-rhee** [reply]
specifically an orthographic projection matrix

**Reply by bobcao3** [reply]
Currently, there's a hacky way to get a perspective projection matrix

```python
camera = ti.ui.Camera()
camera.position(1, 1, 1)  # set camera position
camera.lookat(0, 0, 0)  # set camera lookat
camera.up(0, 1, 0)  # set camera up vector

proj_mtx = camera.get_projection_matrix(aspect_ratio)
```

**Comment by kurt-rhee** [comment]
Also projections of a vector onto another vector

**Comment by Mirebeau** [comment]
The QR decomposition is basic routine that I would use.
(It amounts to orthonormalize a family of vectors)
https://en.wikipedia.org/wiki/QR_decomposition
https://numpy.org/doc/stable/reference/generated/numpy.linalg.qr.html

It is straightforward to implement IF the target matrix has maximal rank. Otherwise you can run into annoying edge cases. 
A robust implementation in the library would be a plus I think.

**Comment by Mirebeau** [comment]
Another request : create a vector with the same dtype as another vector.

Example : u is  a vector with dtype=ti.i8
v = ti.zero(u) # v is a vector with dtype=ti.i32

More generally, it would be great to have u.dtype for a vector/matrix u. (Currently, that only works for a field.)
